pkg/controller/che/che_controller.go

```diff
@@ -574,44 +559,60 @@ func (r *ReconcileChe) Reconcile(request reconcile.Request) (reconcile.Result, e
 	}
 
 	// create Che service and route
-	cheLabels := deploy.GetLabels(instance, util.GetValue(instance.Spec.Server.CheFlavor, deploy.DefaultCheFlavor))
+	serviceStatus := deploy.SyncCheServiceToCluster(instance, clusterAPI)
+	if !tests {
+		if !serviceStatus.Continue {
+			logrus.Info("Waiting on service 'che-host' to be ready")
+			if serviceStatus.Err != nil {
+				logrus.Error(serviceStatus.Err)
+			}
 
-	if _, err := deploy.NewCheService(instance, cheLabels, r); err != nil {
-		return reconcile.Result{}, err
+			return reconcile.Result{Requeue: serviceStatus.Requeue}, serviceStatus.Err
+		}
 	}
+
 	if !isOpenShift {
-		cheIngress := deploy.NewIngress(instance, cheFlavor, "che-host", 8080)
-		if err := r.CreateNewIngress(instance, cheIngress); err != nil {
-			return reconcile.Result{}, err
+		ingressStatus := deploy.SyncIngressToCluster(instance, cheFlavor, "che-host", 8080, clusterAPI)
+		if !tests {
+			if !ingressStatus.Continue {
+				logrus.Info("Waiting on ingress 'che-host' to be ready")
+				if ingressStatus.Err != nil {
+					logrus.Error(ingressStatus.Err)
+				}
+
+				return reconcile.Result{Requeue: ingressStatus.Requeue}, ingressStatus.Err
+			}
 		}
+
 		cheHost := ingressDomain
 		if ingressStrategy == "multi-host" {
 			cheHost = cheFlavor + "-" + instance.Namespace + "." + ingressDomain
 		}
-		if len(instance.Spec.Server.CheHost) == 0 {
+		if instance.Spec.Server.CheHost != cheHost {
 			instance.Spec.Server.CheHost = cheHost
 			if err := r.UpdateCheCRSpec(instance, "CheHost URL", cheHost); err != nil {
 				instance, _ = r.GetCR(request)
 				return reconcile.Result{Requeue: true, RequeueAfter: time.Second * 1}, err
 			}
 		}
 	} else {
-		cheRoute := deploy.NewRoute(instance, cheFlavor, "che-host", 8080)
-		if tlsSupport {
-			cheRoute = deploy.NewTlsRoute(instance, cheFlavor, "che-host", 8080)
-		}
-		if err := r.CreateNewRoute(instance, cheRoute); err != nil {
-			return reconcile.Result{}, err
-		}
-		if len(instance.Spec.Server.CheHost) == 0 {
-			instance.Spec.Server.CheHost = cheRoute.Spec.Host
-			if len(cheRoute.Spec.Host) < 1 {
-				cheRoute := r.GetEffectiveRoute(instance, cheRoute.Name)
-				instance.Spec.Server.CheHost = cheRoute.Spec.Host
+		routeStatus := deploy.SyncRouteToCluster(instance, cheFlavor, "che-host", 8080, clusterAPI)
+		if !tests {
```

Comment:
At some point we should try to get rid of those conditions that fill the implementation with non-tested parts, wdyt ?
Surely in a distinct PR, but I'm just mentioning.

Reply:
BTW the currently refactoring will make unit test implementation on the main logic *much easier*